lib/spack/spack/binary_distribution.py

```diff
@@ -100,436 +79,334 @@ class BinaryCacheIndex(object):
     def __init__(self, cache_root):
         self._index_cache_root = cache_root
 
-        # the key associated with the serialized _local_index_cache
-        self._index_contents_key = "contents.json"
-
         # a FileCache instance storing copies of remote binary cache indices
-        self._index_file_cache = None
-
-        # stores a map of mirror URL to index hash and cache key (index path)
-        self._local_index_cache = None
+        self._index_file_cache = file_cache.FileCache(self._index_cache_root)
 
-        # hashes of remote indices already ingested into the concrete spec
-        # cache (_mirrors_for_spec)
-        self._specs_already_associated = set()
-
-        # mapping from mirror urls to the time.time() of the last index fetch and a bool indicating
-        # whether the fetch succeeded or not.
+        # mapping from mirror urls to the time.time() of the last index fetch.
         self._last_fetch_times = {}
 
-        # _mirrors_for_spec is a dictionary mapping DAG hashes to lists of
-        # entries indicating mirrors where that concrete spec can be found.
-        # Each entry is a dictionary consisting of:
-        #
-        #     - the mirror where the spec is, keyed by ``mirror_url``
-        #     - the concrete spec itself, keyed by ``spec`` (including the
-        #           full hash, since the dag hash may match but we want to
-        #           use the updated source if available)
-        self._mirrors_for_spec = {}
+        # mapping from mirror urls to either:
+        #    The set[str] of DAG hashes available in that mirror, or
+        #    The dict[str, Spec] of concrete specs available indexed by DAG hash.
+        self._mirror_specs = {}
 
-    def _init_local_index_cache(self):
-        if not self._index_file_cache:
-            self._index_file_cache = file_cache.FileCache(self._index_cache_root)
+    @classmethod
+    def _cache_keys_for(cls, mirror_url: str, discriminator: int = 0):
```

Comment:
What's wrong with the mirror name, it's unique.

Reply:
Multiple Spack environments may refer to the same mirror URL via different names, or to different URLs with the same name. They aren't unique enough, so this code works by URL. (It did before too, that trait hasn't changed.)